From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Istvan Zoltan Ballok <istvan.zoltan.ballok@sap.com>
Date: Tue, 10 Dec 2024 13:38:44 +0100
Subject: [PATCH 04/29] Implement the Cooperative Executor

The Cooperative Executor is a scheduler that can be used in unit tests to check
all possible execution paths of concurrently executed functions using some
"cooperative" client APIs, to assert that there is no ordering that would lead
to an unexpected result.

ginkgo  ./pkg/resourcemanager/controller/garbagecollector/race    | less -SNR
go test ./pkg/resourcemanager/controller/garbagecollector/race -v | less -SNR
---
 .../race/cooperative_executor.go              | 290 ++++++++++++++++++
 .../race/cooperative_executor_test.go         | 288 +++++++++++++++++
 .../garbagecollector/race/race_suite_test.go  |  17 +
 3 files changed, 595 insertions(+)
 create mode 100644 pkg/resourcemanager/controller/garbagecollector/race/cooperative_executor.go
 create mode 100644 pkg/resourcemanager/controller/garbagecollector/race/cooperative_executor_test.go
 create mode 100644 pkg/resourcemanager/controller/garbagecollector/race/race_suite_test.go

diff --git a/pkg/resourcemanager/controller/garbagecollector/race/cooperative_executor.go b/pkg/resourcemanager/controller/garbagecollector/race/cooperative_executor.go
new file mode 100644
index 0000000000..883cbaae2e
--- /dev/null
+++ b/pkg/resourcemanager/controller/garbagecollector/race/cooperative_executor.go
@@ -0,0 +1,290 @@
+// SPDX-FileCopyrightText: 2025 SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Package race provides a CooperativeExecutor that can be used in unit tests to
+// check all possible execution paths of concurrently executed functions using some
+// client APIs, to assert that there is no ordering that would lead to an
+// unexpected result.
+//
+// For a more detailed documentation, please see the cooperative_executor\.md file.
+package race
+
+import (
+	"fmt"
+	"sort"
+	"sync"
+	"time"
+)
+
+type CooperativeExecutor interface {
+	RunAllCombinations(setup func(), assertion func() error, fs ...func()) *Result
+	LockAndYield(clientName string, callID int, description string)
+	Release(clientName string, callID int, result string)
+}
+
+type Result struct {
+	NumberOfPathsChecked             int
+	NumberOfPathsWithFailedAssertion int
+	PathsChecked                     string
+	errors                           []error
+}
+
+type LogLevel int
+
+var DEBUG, INFO LogLevel = 0, 1
+
+func NewCooperativeExecutor(logLevel LogLevel) CooperativeExecutor {
+	return &cooperativeExecutor{logLevel: logLevel}
+}
+
+type (
+	cooperativeExecutor struct {
+		logLevel          LogLevel
+		inbox             chan *msg
+		state             state
+		inCriticalSection bool
+		guide             path
+		currentPath       path
+		pathsToCheck      []path
+		checkedPaths      []path
+		assertionResults  []error
+	}
+
+	msg struct {
+		kind        Kind
+		clientName  string
+		callID      int
+		description string
+		result      string
+		proceed     chan struct{}
+	}
+
+	state []*msg
+
+	Kind string
+
+	choice struct {
+		clientName  string
+		callID      int
+		description string
+		result      string
+	}
+
+	path []*choice
+)
+
+const (
+	LockAndYield Kind = "LockAndYield"
+	Release      Kind = "Release"
+)
+
+func (e *cooperativeExecutor) LockAndYield(clientName string, callID int, description string) {
+	m := &msg{
+		kind:        LockAndYield,
+		clientName:  clientName,
+		callID:      callID,
+		description: description,
+		proceed:     make(chan struct{})}
+	e.inbox <- m
+	<-m.proceed
+}
+
+func (e *cooperativeExecutor) Release(clientName string, callID int, result string) {
+	m := &msg{
+		kind:       Release,
+		clientName: clientName,
+		callID:     callID,
+		result:     result}
+	e.inbox <- m
+}
+
+func (e *cooperativeExecutor) run(fs ...func()) {
+	e.inbox = make(chan *msg)
+	var wg sync.WaitGroup
+	for _, f := range fs {
+		wg.Add(1)
+		go func(f func()) {
+			defer wg.Done()
+			f()
+		}(f)
+	}
+
+	var wg2 sync.WaitGroup
+	wg2.Add(1)
+	go func() {
+		defer wg2.Done()
+		timer := time.NewTimer(10 * time.Millisecond)
+		defer timer.Stop()
+		for {
+			select {
+			case msg, ok := <-e.inbox:
+				if !ok {
+					return
+				}
+				e.maintainState(msg)
+				if !timer.Stop() {
+					select {
+					case <-timer.C:
+					default:
+					}
+				}
+				timer.Reset(10 * time.Millisecond)
+			case <-timer.C:
+				e.proceed()
+			}
+		}
+	}()
+
+	wg.Wait()
+	close(e.inbox)
+	wg2.Wait()
+}
+
+func (e *cooperativeExecutor) maintainState(msg *msg) {
+	switch msg.kind {
+	case LockAndYield:
+		e.state = append(e.state, msg)
+	case Release:
+		e.removeStateEntry(msg)
+		e.currentPath[len(e.currentPath)-1].result = msg.result
+		e.inCriticalSection = false
+	}
+}
+
+func (e *cooperativeExecutor) removeStateEntry(msg *msg) {
+	var index int
+	for i, s := range e.state {
+		if s.clientName == msg.clientName && s.callID == msg.callID {
+			index = i
+			break
+		}
+	}
+	e.state = append(e.state[:index], e.state[index+1:]...)
+}
+
+func (e *cooperativeExecutor) proceed() {
+	if e.inCriticalSection || len(e.state) == 0 {
+		return
+	}
+	sort.Slice(e.state, func(i, j int) bool {
+		return e.state[i].clientName < e.state[j].clientName ||
+			(e.state[i].clientName == e.state[j].clientName &&
+				e.state[i].description < e.state[j].description)
+	})
+	if e.logLevel == DEBUG {
+		// fmt.Printf is used for concise multi line text output instead of a regular logger
+		fmt.Printf("Yield point:\n%s", e.state)
+	}
+	var next *msg
+	if len(e.guide) == 0 {
+		// No guided choice, proceed with the first choice and explore the other choices later
+		next = e.state[0]
+		for _, m := range e.state[1:] {
+			e.pathsToCheck = append(
+				e.pathsToCheck,
+				append(
+					append(
+						[]*choice{},
+						e.currentPath...),
+					&choice{
+						clientName:  m.clientName,
+						callID:      m.callID,
+						description: m.description,
+						result:      m.result}))
+		}
+	} else {
+		// Proceed with the guided choice
+		for _, s := range e.state {
+			if e.guide[0].clientName == s.clientName &&
+				e.guide[0].description == s.description {
+				next = s
+				e.guide = e.guide[1:]
+				break
+			}
+		}
+		if next == nil {
+			fmt.Printf("\n  State:\n%s\n", e.state)
+			fmt.Printf("  Guide:\n%s\n", e.guide)
+			panic("Unexpected state: the guided choice does not match any in the current state")
+		}
+	}
+	e.currentPath = append(
+		e.currentPath,
+		&choice{
+			clientName:  next.clientName,
+			callID:      next.callID,
+			description: next.description,
+			result:      next.result})
+	if e.logLevel == DEBUG {
+		fmt.Printf("  Decided to proceed with client: %s/%d %s\n\n", next.clientName, next.callID, next.description)
+	}
+	e.inCriticalSection = true
+	next.proceed <- struct{}{}
+}
+
+func (e *cooperativeExecutor) RunAllCombinations(setup func(), assertion func() error, fs ...func()) *Result {
+	errors := []error{}
+	for {
+		if len(e.pathsToCheck) > 0 {
+			e.guide = e.pathsToCheck[0]
+			e.pathsToCheck = e.pathsToCheck[1:]
+		}
+		time.Sleep(10 * time.Millisecond) // prevent interleaved log output with the test environment that might log concurrently
+		fmt.Printf("\nChecking path #%d. with '%d' guided choices:\n%s\n", len(e.checkedPaths), len(e.guide), e.guide)
+
+		setup()
+		e.run(fs...)
+		err := assertion()
+		fmt.Printf("\nExplored execution path:\n%s\n", e.currentPath)
+		fmt.Printf("  assertion error: %v\n", err)
+
+		if e.logLevel == DEBUG {
+			fmt.Printf("\nPaths to check later: %d\n\n", len(e.pathsToCheck))
+			for i, path := range e.pathsToCheck {
+				fmt.Printf("  Path %d:\n%s\n", i+1, path)
+			}
+		}
+		e.assertionResults = append(e.assertionResults, err)
+		if err != nil {
+			errors = append(errors, err)
+		}
+		e.checkedPaths = append(e.checkedPaths, e.currentPath)
+		e.currentPath = path{}
+		if len(e.pathsToCheck) == 0 {
+			break
+		}
+	}
+	fmt.Printf("\n\nOut of '%d' checked paths, '%d' had failed assertions.\n", len(e.checkedPaths), len(errors))
+	return &Result{
+		NumberOfPathsChecked:             len(e.checkedPaths),
+		NumberOfPathsWithFailedAssertion: len(errors),
+		PathsChecked:                     e.pathsChecked(),
+		errors:                           errors,
+	}
+}
+
+func (e *cooperativeExecutor) pathsChecked() string {
+	result := ""
+	for i, path := range e.checkedPaths {
+		result += fmt.Sprintf("Path %d:\n%s", i+1, path)
+		result += fmt.Sprintf("  assertion error: %v\n", e.assertionResults[i])
+	}
+	return result
+}
+
+func (p path) String() string {
+	result := ""
+	for _, choice := range p {
+		result += fmt.Sprintf("- %s/%d %s -> %s\n", choice.clientName, choice.callID, choice.description, choice.result)
+	}
+	return result
+}
+
+func (s state) String() string {
+	result := ""
+	for _, m := range s {
+		result += fmt.Sprintf("- %s/%d %s\n", m.clientName, m.callID, m.description)
+	}
+	return result
+}
diff --git a/pkg/resourcemanager/controller/garbagecollector/race/cooperative_executor_test.go b/pkg/resourcemanager/controller/garbagecollector/race/cooperative_executor_test.go
new file mode 100644
index 0000000000..7757f679db
--- /dev/null
+++ b/pkg/resourcemanager/controller/garbagecollector/race/cooperative_executor_test.go
@@ -0,0 +1,288 @@
+// SPDX-FileCopyrightText: 2025 SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package race_test
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/gardener/gardener/pkg/resourcemanager/controller/garbagecollector/race"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+)
+
+type SimpleClient interface {
+	Get() (value int)
+	Set(value int)
+	GetWithVersion() (value, version int)
+	SetWithVersion(value, version int) error
+}
+
+func NewSimpleClient() SimpleClient {
+	return &simpleClient{}
+}
+
+type simpleClient struct {
+	mu      sync.Mutex
+	version int
+	value   int
+}
+
+func (c *simpleClient) Get() (value int) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.value
+}
+
+func (c *simpleClient) Set(value int) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.set(value)
+}
+
+func (c *simpleClient) set(value int) {
+	c.value = value
+	c.version++
+}
+
+func (c *simpleClient) GetWithVersion() (value, version int) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.value, c.version
+}
+
+func (c *simpleClient) SetWithVersion(value, version int) (err error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.version != version {
+		return fmt.Errorf("version mismatch: expected %d, got %d", c.version, version)
+	}
+	c.set(value)
+	return nil
+}
+
+func NewCooperativeSimpleClient(clientName string, simpleClient SimpleClient, executor race.CooperativeExecutor) SimpleClient {
+	return &cooperativeSimpleClient{
+		clientName:   clientName,
+		simpleClient: simpleClient,
+		executor:     executor,
+	}
+}
+
+type cooperativeSimpleClient struct {
+	clientName   string
+	simpleClient SimpleClient
+	executor     race.CooperativeExecutor
+	mu           sync.Mutex
+	callID       int
+}
+
+func (cc *cooperativeSimpleClient) nextCallID() int {
+	cc.mu.Lock()
+	defer cc.mu.Unlock()
+	callID := cc.callID
+	cc.callID++
+	return callID
+}
+
+func (cc *cooperativeSimpleClient) Get() int {
+	callID := cc.nextCallID()
+	cc.executor.LockAndYield(cc.clientName, callID, "Get()")
+	value := cc.simpleClient.Get()
+	cc.executor.Release(cc.clientName, callID, fmt.Sprintf("value=%d", value))
+	return value
+}
+
+func (cc *cooperativeSimpleClient) Set(value int) {
+	nextCallID := cc.nextCallID()
+	cc.executor.LockAndYield(cc.clientName, nextCallID, fmt.Sprintf("Set(%d)", value))
+	cc.simpleClient.Set(value)
+	cc.executor.Release(cc.clientName, nextCallID, fmt.Sprintf("[value=%d]", cc.simpleClient.Get()))
+}
+
+func (cc *cooperativeSimpleClient) GetWithVersion() (value, version int) {
+	nextCallID := cc.nextCallID()
+	cc.executor.LockAndYield(cc.clientName, nextCallID, "GetWithVersion()")
+	value, version = cc.simpleClient.GetWithVersion()
+	cc.executor.Release(cc.clientName, nextCallID, fmt.Sprintf("value=%d, version=%d", value, version))
+	return
+}
+
+func (cc *cooperativeSimpleClient) SetWithVersion(value, version int) error {
+	nextCallID := cc.nextCallID()
+	cc.executor.LockAndYield(cc.clientName, nextCallID, fmt.Sprintf("SetWithVersion(%d, %d)", value, version))
+	err := cc.simpleClient.SetWithVersion(value, version)
+	cc.executor.Release(cc.clientName, nextCallID, fmt.Sprintf("[value=%d], error=%v", cc.simpleClient.Get(), err))
+	return err
+}
+
+var _ = Describe("Cooperative Executor", func() {
+
+	It("should detect possible race conditions when two actors try to increment a shared value without any synchronization mechanism", func() {
+		executor := race.NewCooperativeExecutor(race.DEBUG)
+		var simpleClient, clientA, clientB SimpleClient
+
+		setup := func() {
+			simpleClient = NewSimpleClient()
+			clientA = NewCooperativeSimpleClient("A", simpleClient, executor)
+			clientB = NewCooperativeSimpleClient("B", simpleClient, executor)
+		}
+
+		assert := func() error {
+			if simpleClient.Get() != 2 {
+				return fmt.Errorf("expected value==2, got value==%d", simpleClient.Get())
+			}
+			return nil
+		}
+
+		result := executor.RunAllCombinations(
+			setup,
+			assert,
+			func() {
+				value := clientA.Get()
+				clientA.Set(value + 1)
+			},
+			func() {
+				value := clientB.Get()
+				clientB.Set(value + 1)
+			},
+		)
+		Expect(result.NumberOfPathsWithFailedAssertion).To(Equal(4))
+		Expect(result.NumberOfPathsChecked).To(Equal(6))
+		Expect(strings.TrimSpace(result.PathsChecked)).To(Equal(strings.TrimSpace(`
+
+Path 1:
+- A/0 Get() -> value=0
+- A/1 Set(1) -> [value=1]
+- B/0 Get() -> value=1
+- B/1 Set(2) -> [value=2]
+  assertion error: <nil>
+Path 2:
+- B/0 Get() -> value=0
+- A/0 Get() -> value=0
+- A/1 Set(1) -> [value=1]
+- B/1 Set(1) -> [value=1]
+  assertion error: expected value==2, got value==1
+Path 3:
+- A/0 Get() -> value=0
+- B/0 Get() -> value=0
+- A/1 Set(1) -> [value=1]
+- B/1 Set(1) -> [value=1]
+  assertion error: expected value==2, got value==1
+Path 4:
+- B/0 Get() -> value=0
+- B/1 Set(1) -> [value=1]
+- A/0 Get() -> value=1
+- A/1 Set(2) -> [value=2]
+  assertion error: <nil>
+Path 5:
+- B/0 Get() -> value=0
+- A/0 Get() -> value=0
+- B/1 Set(1) -> [value=1]
+- A/1 Set(1) -> [value=1]
+  assertion error: expected value==2, got value==1
+Path 6:
+- A/0 Get() -> value=0
+- B/0 Get() -> value=0
+- B/1 Set(1) -> [value=1]
+- A/1 Set(1) -> [value=1]
+  assertion error: expected value==2, got value==1
+
+		`)), "Here is the actual value of PathsChecked in full length. Copy this and update the string literal in the unit test's source code: \n\n%s\n",
+			result.PathsChecked)
+	})
+
+	It("should detect no race conditions when the actors properly use optimistic locking", func() {
+		executor := race.NewCooperativeExecutor(race.DEBUG)
+		var simpleClient, clientA, clientB SimpleClient
+
+		setup := func() {
+			simpleClient = NewSimpleClient()
+			clientA = NewCooperativeSimpleClient("A", simpleClient, executor)
+			clientB = NewCooperativeSimpleClient("B", simpleClient, executor)
+		}
+
+		assert := func() error {
+			if simpleClient.Get() != 2 {
+				return fmt.Errorf("expected value==2, got value==%d", simpleClient.Get())
+			}
+			return nil
+		}
+
+		result := executor.RunAllCombinations(
+			setup,
+			assert,
+			func() {
+				for {
+					value, version := clientA.GetWithVersion()
+					if clientA.SetWithVersion(value+1, version) == nil {
+						break
+					}
+				}
+			},
+			func() {
+				for {
+					value, version := clientB.GetWithVersion()
+					if clientB.SetWithVersion(value+1, version) == nil {
+						break
+					}
+				}
+			},
+		)
+		Expect(result.NumberOfPathsWithFailedAssertion).To(Equal(0))
+		Expect(result.NumberOfPathsChecked).To(Equal(6))
+
+		Expect(strings.TrimSpace(result.PathsChecked)).To(Equal(strings.TrimSpace(`
+
+Path 1:
+- A/0 GetWithVersion() -> value=0, version=0
+- A/1 SetWithVersion(1, 0) -> [value=1], error=<nil>
+- B/0 GetWithVersion() -> value=1, version=1
+- B/1 SetWithVersion(2, 1) -> [value=2], error=<nil>
+  assertion error: <nil>
+Path 2:
+- B/0 GetWithVersion() -> value=0, version=0
+- A/0 GetWithVersion() -> value=0, version=0
+- A/1 SetWithVersion(1, 0) -> [value=1], error=<nil>
+- B/1 SetWithVersion(1, 0) -> [value=1], error=version mismatch: expected 1, got 0
+- B/2 GetWithVersion() -> value=1, version=1
+- B/3 SetWithVersion(2, 1) -> [value=2], error=<nil>
+  assertion error: <nil>
+Path 3:
+- A/0 GetWithVersion() -> value=0, version=0
+- B/0 GetWithVersion() -> value=0, version=0
+- A/1 SetWithVersion(1, 0) -> [value=1], error=<nil>
+- B/1 SetWithVersion(1, 0) -> [value=1], error=version mismatch: expected 1, got 0
+- B/2 GetWithVersion() -> value=1, version=1
+- B/3 SetWithVersion(2, 1) -> [value=2], error=<nil>
+  assertion error: <nil>
+Path 4:
+- B/0 GetWithVersion() -> value=0, version=0
+- B/1 SetWithVersion(1, 0) -> [value=1], error=<nil>
+- A/0 GetWithVersion() -> value=1, version=1
+- A/1 SetWithVersion(2, 1) -> [value=2], error=<nil>
+  assertion error: <nil>
+Path 5:
+- B/0 GetWithVersion() -> value=0, version=0
+- A/0 GetWithVersion() -> value=0, version=0
+- B/1 SetWithVersion(1, 0) -> [value=1], error=<nil>
+- A/1 SetWithVersion(1, 0) -> [value=1], error=version mismatch: expected 1, got 0
+- A/2 GetWithVersion() -> value=1, version=1
+- A/3 SetWithVersion(2, 1) -> [value=2], error=<nil>
+  assertion error: <nil>
+Path 6:
+- A/0 GetWithVersion() -> value=0, version=0
+- B/0 GetWithVersion() -> value=0, version=0
+- B/1 SetWithVersion(1, 0) -> [value=1], error=<nil>
+- A/1 SetWithVersion(1, 0) -> [value=1], error=version mismatch: expected 1, got 0
+- A/2 GetWithVersion() -> value=1, version=1
+- A/3 SetWithVersion(2, 1) -> [value=2], error=<nil>
+  assertion error: <nil>
+
+		`)), "Here is the actual value of PathsChecked in full length. Copy this and update the string literal in the unit test's source code: \n\n%s\n",
+			result.PathsChecked)
+	})
+})
diff --git a/pkg/resourcemanager/controller/garbagecollector/race/race_suite_test.go b/pkg/resourcemanager/controller/garbagecollector/race/race_suite_test.go
new file mode 100644
index 0000000000..bc211748cd
--- /dev/null
+++ b/pkg/resourcemanager/controller/garbagecollector/race/race_suite_test.go
@@ -0,0 +1,17 @@
+// SPDX-FileCopyrightText: 2025 SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package race_test
+
+import (
+	"testing"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+)
+
+func TestGarbageCollector(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "ResourceManager Controller GarbageCollector Race Suite")
+}
